### PR TITLE
DB-10768 Make Input Schema Configurable

### PR DESCRIPTION
**Pull request by jpanko1**
Input Schema Example:
For the 4th argument to the KafkaReader main function, pass a DDL string like
"ID STRING NOT NULL, LOCATION STRING, TEMPERATURE FLOAT, HUMIDITY FLOAT, TM TIMESTAMP"

### CODE REVIEW

File: spark2.4/build.sbt [hunk 1]
```diff
@@ -7,7 +7,7 @@ scalaVersion := "2.11.12"
 
 //val spliceVersion = "2.8.0.1945"
 //val spliceVersion = "3.1.0.1967-SNAPSHOT"
-val spliceVersion = "3.2.0.1980-SNAPSHOT"
+val spliceVersion = "3.2.0.1985-SNAPSHOT"
```
Comment (jaceklaskowski) [inline spark2.4/build.sbt:10]: While you're at it, why not to remove the above commented-out versions too?

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 2]
```diff
@@ -45,23 +47,33 @@ object KafkaReader {
 //
 //    val metricsProducer = new KafkaProducer[Integer, Long](props)
 //    val metricsTopic = "ssds-metrics"
-
+    
+    // Create schema from ddl string like
+    //    "ID STRING NOT NULL, LOCATION STRING, TEMPERATURE DOUBLE, HUMIDITY DOUBLE, TM TIMESTAMP"
+    var schema = new StructType
+    schemaDDL.split(",").foreach( s => {
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:54]:
```
foreach { s => 
  ...
}
```

When you start with `{` you can avoid `{` after `=>`.

Comment (jaceklaskowski) [reply spark2.4/src/main/scala/KafkaReader.scala:54]: You could also use [StructType.fromDDL](https://github.com/apache/spark/blob/290a528bff7bcb449714c1c6f1885bd0f804358d/sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala#L539).

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 3]
```diff
@@ -45,23 +47,33 @@ object KafkaReader {
...
+    // Keep for reference -- Performance test (first five fields)
+//    val schema = StructType(
+//      StructField("ID", LongType, false) ::
+//      StructField("PAYLOAD", StringType, true) ::
+//      StructField("SRC_SERVER", StringType, false) ::
+//      StructField("SRC_THREAD", LongType, true) ::
+//      StructField("TM_GENERATED", LongType, false) :: Nil)
+////      StructField("PTN_NSDS", IntegerType, true) ::
+////      StructField("TM_NSDS", LongType, true) :: Nil)
+////      StructField("TM_EXT_KAFKA", LongType, true) :: Nil)
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:76]: Please remove all the lines (they're in git anyway).

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 4]
```diff
@@ -102,11 +113,17 @@ object KafkaReader {
       reader.option("kafka.client.id", s"$group-$clientId")  // probably should use uuid instead of user input TODO
     }
     
-    val values = reader
-      .load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )) as "data", col("timestamp") cast "long" as "TM_EXT_KAFKA")
-      .select("data.*", "TM_EXT_KAFKA")
-      .withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)
-      .withColumn("TM_SSDS", unix_timestamp * 1000 )
+    val values = if( useFlowMarkers ) {
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:116]: nit: "replace" spaces to be `if (useFlowMarkers) {`

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 5]
```diff
@@ -102,11 +113,17 @@ object KafkaReader {
...
+      reader
+        .load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )) as "data", col("timestamp") cast "long" as "TM_EXT_KAFKA")
+        .select("data.*", "TM_EXT_KAFKA")
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:119]: Why not to "join" this `select` with the above `select`? Why two?

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 6]
```diff
@@ -102,11 +113,17 @@ object KafkaReader {
...
+        .withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:120]: Why not to "merge" it with the above `select`? It's confusing.

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 7]
```diff
@@ -102,11 +113,17 @@ object KafkaReader {
...
+        .withColumn("TM_SSDS", unix_timestamp * 1000 )
+    } else {
+      reader
+        .load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )) as "data")
+        .select("data.*")
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:125]: Again, I don't understand why to `from_json` and `Map(...) as "data"` for `select("data.*")`? I seem to be failing to find something important here.

File: spark2.4/src/main/scala/KafkaReader.scala [hunk 8]
```diff
@@ -157,6 +176,7 @@ object KafkaReader {
 //          spliceUrl,
 //          spliceKafkaServers,
 //          spliceKafkaPartitions,
+//          useFlowMarkers,
```
Comment (jaceklaskowski) [inline spark2.4/src/main/scala/KafkaReader.scala:179]: Remove the whole commented-out block.

**Review by jaceklaskowski — CHANGES_REQUESTED** [review]

File: spark2.4/build.sbt (hunk 1 above)
Comment (jpanko1) [reply spark2.4/build.sbt:8]: Removed in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/27/commits/f67f17908f97ceeaddf328c82576ad815f7e3afa

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 2 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala]: Changed foreach format in commit https://github.com/splicemachine/splice-machine-spark-connector/pull/27/commits/43f72a64c5a3c29c4609801af6e8f428e4fde6aa

I do prefer using functions from a library and tried to use StructType.fromDDL, but it didn’t support "NOT NULL".
`scala> StructType.fromDDL("ID STRING NOT NULL, LOCATION STRING, TEMPERATURE DOUBLE, HUMIDITY DOUBLE, TM TIMESTAMP")`
`org.apache.spark.sql.catalyst.parser.ParseException:
mismatched input 'NOT' expecting <EOF>(line 1, pos 10)
`

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 3 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala]: Removed in commit https://github.com/splicemachine/splice-machine-spark-connector/commit/43f72a64c5a3c29c4609801af6e8f428e4fde6aa

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 4 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala]: Updated in commit https://github.com/splicemachine/splice-machine-spark-connector/commit/43f72a64c5a3c29c4609801af6e8f428e4fde6aa

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 6 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala]: Updated in commit https://github.com/splicemachine/splice-machine-spark-connector/commit/43f72a64c5a3c29c4609801af6e8f428e4fde6aa

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 5 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala]: I’m open to suggestions, but I believe the code is in line with the examples shown at:
https://databricks.com/blog/2017/04/26/processing-data-in-apache-kafka-with-structured-streaming-in-apache-spark-2-2.html
https://sparkbyexamples.com/spark/spark-streaming-with-kafka/
https://stackoverflow.com/questions/42506801/how-to-use-from-json-with-kafka-connect-0-10-and-spark-structured-streaming
https://jaceklaskowski.github.io/mastering-spark-sql-book/spark-sql-functions-collection/

I also just tried some alternatives like the following, and they all resulted in the exception that follows.
`.load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )), col("timestamp") cast "long" as "TM_EXT_KAFKA")`
`.withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)`
`.withColumn("TM_SSDS", unix_timestamp * 1000 )`

`Exception in thread "main" java.lang.IllegalArgumentException: Can't get JDBC type for struct<ID:bigint,PAYLOAD:string,SRC_SERVER:string,SRC_THREAD:bigint,TM_GENERATED:bigint>`

The data needs to be pulled out of the struct.

Trying explode resulted in a different exception.

`.load.select(explode(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" ))), col("timestamp") cast "long" as "TM_EXT_KAFKA")`
`.withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)`
`.withColumn("TM_SSDS", unix_timestamp * 1000 )`

`Exception in thread "main" org.apache.spark.sql.AnalysisException: cannot resolve 'explode(jsontostructs(CAST(`value` AS STRING)))' due to data type mismatch: input to function explode should be array or map type, not struct`

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 7 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala:105]: I believe this is the same issue as two comments above.  If there's something else, please let me know.

File: spark2.4/src/main/scala/KafkaReader.scala (hunk 8 above)
Comment (jpanko1) [reply spark2.4/src/main/scala/KafkaReader.scala:159]: Entered Jira [DB-10873](https://splicemachine.atlassian.net/browse/DB-10873) to uncomment the block and to add an input param to choose between using ParallelLoaderInserter or the Loader + Inserter pair in the preceding lines.  They are alternative implementations for inserting the data in SpliceMachine.

**Review by martinrupp — COMMENTED** [review]
how can the input schema be configured now? can you provide a small example in the description?

**Review by arnaud-splice — APPROVED** [review]

**Comment by jpanko1** [comment]
> how can the input schema be configured now? can you provide a small example in the description?

@martinrupp Added input schema example to the description.

**Comment by martinrupp** [comment]
i don't really see a test for this, can we add one?

**Review by bklo94 — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**